dissect/target/tools/build_magic.py

```diff
@@ -0,0 +1,210 @@
+            if ":" in (offset := magic_match.get("offset", "0")):
+                offset = offset.split(":")[0]
+            try:
+                offset = int(offset)
+            except ValueError:
+                pass
+
+            if (offset, value) in magics_seen:
+                print(f"Duplicate magic value for type {mime_type_name!r}: offset={offset!r} value={value!r}")
+                continue
+
+            if len(value) < 2:
+                print(
+                    f"Magic value too short to be usable for type {mime_type_name!r}: offset={offset!r} value={value!r}"
+                )
+                continue
+
+            magics_seen.add((offset, value))
+            magics.append({"offset": offset, "value": value})
+
+        patterns = [g.get("pattern") for g in mime_type.findall("mime:glob", ns)]
+
+        if not patterns and not magics:
+            print(f"No usable definitions for {mime_type_name!r}")
+            continue
+
+        types.append(
+            {
+                "type": mime_type_name,
+                "name": comment,
+                "pattern": patterns,
+                "magic": magics,
+            }
+        )
+
+    print(f"Parsed {len(types)} types")
+
+    return TEMPLATE.format(timestamp=datetime.now(tz=timezone.utc), count=len(types), hash=hash, types=types)
+
+
+def convert_value(type: str, value: str) -> bytes:
+    """Convert a value to bytes."""
+
+    # Regular string conversion to bytes
+    # e.g. "PK\\003\\004"       -> b"PK\x03\x04"
+    # e.g. "\\x00\\x01"         -> b"\x00\x01"
+    # e.g. "\x89PNG\r\n\x1A\n"  -> b"\x89PNG\r\n\x1a\n"
+    if type == "string":
+        if "\\x" in value or "\\0" in value:
+            return value.encode("latin1").decode("unicode-escape").encode("latin1")
+        return value.encode("latin1")
+
+    # Integer conversion
+    # e.g. "0x9501"     -> b"\x95\x01"
+    # e.g. "0xa1b2c3d4" -> b"\xa1\xb2\xc3\xd4"
+    if type in ("byte", "little16", "little32", "big16", "big32", "host16", "host32"):
+        return bytes.fromhex(value[2:] if value[:2] == "0x" else value)
+
+    raise ValueError(f"Unknown type {type!r}")
+
+
+@catch_sigpipe
+def main() -> int:
+    help_formatter = argparse.ArgumentDefaultsHelpFormatter
+    parser = argparse.ArgumentParser(
+        description="target-build-magic",
+        fromfile_prefix_chars="@",
+        formatter_class=help_formatter,
+    )
+    parser.add_argument("path", metavar="PATH", type=Path, help="Path to freedesktop.org.xml.in file")
+    parser.add_argument("--output", "-o", type=Path, help="Optional output path for generated types document")
+    parser.add_argument("--format", action="store_true", default=False, help="Format output using ruff")
+    args = parser.parse_args()
+
+    if not args.path.is_file():
+        print(f"Provided file does not exist: {args.path}")
+        return 1
+
+    output = (args.output or Path("dissect/target/helpers/magic/mimetypes/freedesktop.py")).resolve()
```

Comment:
This assumes that the tool is executed from the root of the repository. While this is a "development" tool, maybe a one line sanity check of that would be nice to add. I.e. checking that the target directory/path exists, and printing are you sure you're in the repository root. Or basing it on the path of `__file__`.